### ISSUE TITLE
Result: We want better diffs

### ISSUE THREAD

**Issue by sils1297** [coala-analyzer/coala#1021]
1. We probably want to allow to pass a function that gets some parameters and *generates* a diff. Useful for e.g. variable renaming, let the user choose the proper name.
2. On occasion there might be several ways to fix a problem, so we may want to provide alternative diffs/generation functions.

**Comment by AbdealiJK** [comment]
I didn't understand this much. could you elaborate on what exactly it means ?
Are you saying you want helper functions that do basic things you'd want with diffs ?

**Comment by sils1297** [comment]
@AbdealiJK no. I want two things:

1. Sometimes you need a value from the user to generate a patch. Say you have a bear that detects badly named variables, it would be easy to generate a diff with that bear *if* you know what the new variable name should be which you don't want to determine automatically. A solution would be to let bears pass a function or something like that, which generates a diff out of some user input, i.e. ask him "what would be a good variable name for that?" and then rename it accordingly.
2. Some bears might have multiple solutions for the same problem. Say LanguageToolBear detects a spelling error and suggests multiple fixes. Currently we have to choose one of them automatically, would be nice to offer all of them to the user in this case.

**Comment by Makman2** [comment]
Btw @sils1297 we have now unified diff, you want it still better? Highlighting works good and I have no problems currently to understand a patch^^

**Comment by sils1297** [comment]
Why does nobody understand what I mean... can you read my last comment or
is it so confusing?

2015-12-06 19:46 GMT+01:00 Mischa Krüger <notifications@github.com>:

**Comment by Makman2** [comment]
ah okay yeah I've misunderstood, so you want a more interactive behaviour, right?

**Comment by sils1297** [comment]
Yeah:

```
|   2| user_name = 'sils1297'
|   3| login(user_name)
|    | [INFO] ConstantsBear:
|    | This variable is a constant and should be capitalized.
|    | The following actions are applicable to this result:
|    |  0: Apply no further actions.
|    |  1: Open the affected file(s) in an editor.
|    |  2: Apply the patch ('capitalize').
|    |  3: Apply the interactive patch ('rename').
|    |  4: Show the patch ('capitalize').
|    |  5: Show the interactive patch ('rename').
|    | Please enter the number of the action you want to execute. 
```

Now if I choose 3 or 5 I will be asked for a new name and a diff will be generated with that info.

**Comment by sils** [comment]
"Interactive patches" might actually be really simple, just let the diff have placeholders using pythons str.format method :) Then the bear can use those and declare variables and we don't even need to distinguish between an interactive and a simple patch.

**Comment by Makman2** [comment]
@sils where do the placeholders are exactly?

**Comment by sils** [comment]
well you'd do a `diff.change_line(6, 'int {new_variable}=0;')` and the diff would accept kwargs so we can instanciate the diff and then apply it

**Comment by Makman2** [comment]
we need to have an option to change the style for such format placeholders^^

anyway, I think not all patches for a certain kind of problem have the same structure. Having some more generic approach may be better^^

**Comment by etotientz** [comment]
@sils so by diff u mean that if user inputs 3 or 5 ,bear will provide a list of acceptable variable names?

**Comment by Makman2** [comment]
you press "apply patch" and coala asks you for inputting some values which will be used to generate the final patch

**Comment by satwikkansal** [comment]
Should we also allow users to modify the diffs proposed by the bears? (in case they don't like some particular part of the fix proposed by the bear and want to do some minor changes in it)

**Comment by Makman2** [comment]
If they want to do minor changes, they shall open their IDE or other environment they develop their project on imo ;)

**Comment by Nosferatul** [comment]
@Makman2 input values should depends on what bears user used, right? Also, some of them could be by default proposed(shouldn't depend on bears)

**Comment by Makman2** [comment]
>  input values should depend on what bears user used, right?

sorry don't understand what you mean.

> Also, some of them could be by default proposed(shouldn't depend on bears)

Yeah definitely :)

**Comment by Nosferatul** [comment]
we will generate different input values for (e.g) SpaceConsistencyBear and LineLengthBear? that's what I asked :)

**Comment by Makman2** [comment]
that would also be possible. But we also want some user interaction so that you can e.g. rename a variable like you want.

**Comment by Nosferatul** [comment]
I am thinking about a profile for a user. Let's say we implement that parameters( e.g variable renaming) and user A wants to use that variable everywhere when he run coala -> he makes a profile let's say profileA.
It will be like a coafile(same syntax) but it will work in every directory(depending on permissions) with or without .coafile in it.

**Comment by Makman2** [comment]
Phew that sounds complicated. Just look at https://github.com/coala/coala/issues/1021#issuecomment-170033355 :)

**Comment by Makman2** [comment]
And when for example selecting `3: Apply the interactive patch ('rename')`, the user will be prompted something like this:
```
Please type the value for 'variable_name':
```
and he types the name that is going to be inserted into the diff wherever a placeholder is defined there.

**Comment by satwikkansal** [comment]
I'm wondering if it's possible to extend this templated diff behavior to external bears.

**Comment by satwikkansal** [comment]
> Should we also allow users to modify the diffs proposed by the bears? (in case they don't like some particular part of the fix proposed by the bear and want to do some minor changes in it)

> If they want to do minor changes, they shall open their IDE or other environment they develop their project on imo ;)

@Makman2 I think this behavior can be helpful in special cases 
Example, when there is a conflict b/w multiple bears because they propose different fixes for the same line, we can simply show all the `result_message` strings to the user, and let the user fix the line in the CLI itself.

**Comment by Makman2** [comment]
Filed https://github.com/coala/coala/issues/4016 for that. Kinda like the idea...

**Comment by satwikkansal** [comment]
> The idea is to use simple placeholders like python uses for strings: Please insert {number} here
So in fact tools would be able to use that, though I doubt someone does... So only choice are upstream changes.

I see, And for "Multiple patches" thing, we'll need to define another `output_format`, right?

One possible approach for the multiple patches functionality that I can think of is : 
After analyzing a piece of code, the bear detects a problem and generates a list of suggestions for that source-range (say the num of suggestions are n). So now we make n `Diff` objects for that source-range. 
Now, if there are k problems detected and m suggestions for each ⇒ k*m `Diff` objects (Assuming all the suggestions for the problems are independent of each other). 
And if the bear produces k different output files corresponding to the analyzed file, the number of `Diff` objects reduce to k. 

If this approach seems okay, how should we create an `output_format` for this behavior so that External Bears can use it?

(or an alternative approach --> We don't really create an output format for this case. Maybe use `regex` based approach to extract multiple results from the linter's output and then do it the same way as we'll do it for the Native bears)

**Comment by Makman2** [comment]
> I see, And for "Multiple patches" thing, we'll need to define another output_format, right?

Oh yeah `@linter` needs to recognize those... phew... But I think we shouldn't do it now, no external tool does support that now, let's care for that later.

> One possible approach for the multiple patches functionality that I can think of is :
After analyzing a piece of code, the bear detects a problem and generates a list of suggestions for that source-range (say the num of suggestions are n). So now we make n Diff objects for that source-range.
Now, if there are k problems detected and m suggestions for each ⇒ k*m Diff objects (Assuming all the suggestions for the problems are independent of each other).
And if the bear produces k different output files corresponding to the analyzed file, the number of Diff objects reduce to k.

See https://github.com/coala/coala/issues/3650 to support this :)

Though the idea is that the user inputs a value to generate a patch, that means we have possibly arbitrary many suggestions, m -> infinity. That's sadly not coverable by https://github.com/coala/coala/issues/3650

**Comment by Vedarth** [comment]
I would like to work on this. Kindly assign me.

**Comment by Naveenaidu** [comment]
@Makman2  I would like to work on this issue. Can i please be assigned to it.

**Comment by Vedarth** [comment]
@satwikkansal worked on https://github.com/coala/coala/issues/3650 which is related to this issue. Can I use his approach to solve this issue? I think it can be solved that way.

**Comment by Makman2** [comment]
Yeah, however this might be quite hard. Also note that we might want to implement the mentioned action system as part of this GSoC 2018.

**Comment by Vedarth** [comment]
@Makman2 I am trying to use it for improve diff handling project of GSoC.

**Comment by Vedarth** [comment]
As it has been almost solved by a pr I suppose we can start working on it right now, I have gone through the discussions and codebase too many times now and I believe I can implement it, also it will help me in preparing the proposal too. Also I had a doubt too, how do you suppose we should implement it for multiple bears. Someone mentioned that we are planning on creating multiple patches for multiple bears at the same time too. Will adding result.next_action field be enough? Because bears run in parallel. Also when number of suggestion become very large 

> Though the idea is that the user inputs a value to generate a patch, that means we have possibly arbitrary many suggestions, m -> infinity. That's sadly not coverable by #3650

So there is still room for improvement. What I am thinking is implement what we already have. And then think for a better solution. 

Coming to the main issue project description for improve diff handling expects us to create a bear which uses multiple patches. So it will be very convenient if we atleast have one generic solution for doing so, before the start of GSoC.

**Comment by Vedarth** [comment]
Don't you think after implementing modified-file coordinated system it will become easier to implement this? I was planning to modify @satwikkansal after implementing modified-file coordinated system.
